implementations/rust/ockam/ockam_api/src/cli_state/traits.rs

```diff
@@ -0,0 +1,123 @@
+use crate::cli_state::{file_stem, CliState, CliStateError};
+use ockam_core::async_trait;
+use serde::{Deserialize, Serialize};
+use std::path::{Path, PathBuf};
+
+use super::Result;
+
+/// Represents the directory of a type of state and contains
+/// all the data related to that type.
+#[async_trait]
+pub trait StateTrait: Sized {
+    type ItemDir: StateItemDirTrait;
+    type ItemConfig: StateItemConfigTrait + Serialize + for<'a> Deserialize<'a>;
+
+    fn new(dir: PathBuf) -> Self;
+    fn default_filename() -> &'static str;
+    fn build_dir(root_path: &Path) -> PathBuf;
+    fn has_data_dir() -> bool;
+
+    #[allow(clippy::new_ret_no_self)]
+    fn load(root_path: &Path) -> Result<Self> {
+        let dir = Self::build_dir(root_path);
+        if Self::has_data_dir() {
+            std::fs::create_dir_all(dir.join("data"))?;
+        } else {
+            std::fs::create_dir_all(&dir)?;
+        }
+        Ok(Self::new(dir))
+    }
+
+    fn dir(&self) -> &PathBuf;
+    fn dir_as_string(&self) -> String {
+        self.dir().to_string_lossy().to_string()
+    }
+
+    fn path(&self, name: &str) -> PathBuf {
+        self.dir().join(format!("{name}.json"))
+    }
+
+    async fn create(&self, name: &str, config: Self::ItemConfig) -> Result<Self::ItemDir>;
```

Comment:
I have some small concerns on the need for async here and in `delete` fns. Especially when we very rarely use our async calls for any concurrent performance gains.

It has impact on our usage the state within clap parsers, and other implementations.

Reply:
There are some cases where we need the async. Maybe we could have `create` and `create_async` to use async just when necessary?

Reply:
perhaps, I didn't realizes there were some cases where we needed the async here.

We can keep it as is then, and then update to add a sync `fn` when we need it.